### PR TITLE
No longer store FileRefs in most of trees, keep them in MethodDef and ClassDefs

### PR DESCRIPTION
**Pull request by DarkDimius**
### Motivation
We want to allow more bits in FileRefs and this means that Loc's will no longer fit in 64 bits. This has substantial performance impact so, we're moving most code _away_ from storing locs to only store LocOffsets

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by DarkDimius** [comment]
cc @jvilk-stripe I expect this to fix memory usage and perf regression

File: core/Loc.h [hunk 1]
```diff
@@ -23,24 +23,32 @@ struct LocOffsets {
...
-} __attribute__((packed, aligned(1)));
-CheckSize(LocOffsets, 6, 1);
+    static LocOffsets none() {
+        return LocOffsets{INVALID_POS_LOC, INVALID_POS_LOC};
+    }
+    LocOffsets join(LocOffsets other) const;
+    // For a given Loc, returns a zero-length version that starts at the same location.
+    LocOffsets copyWithZeroLength() const {
+        return LocOffsets{beginPos(), beginPos()};
+    }
```
Comment (jez) [inline core/Loc.h:33]: Does this function still make sense here? Maybe it should move to `Loc`, or be inlined into the places that were using it?

Comment (DarkDimius) [reply core/Loc.h:33]: Happy either way. Do you want it done as part of this PR?

File: ast/desugar/Desugar.cc [hunk 2]
```diff
@@ -166,51 +166,52 @@ unique_ptr<Expression> validateRBIBody(DesugarContext dctx, unique_ptr<Expressio
     if (!body->loc.exists()) {
         return body;
     }
-    if (!body->loc.file().data(dctx.ctx).isRBI()) {
+    auto loc = core::Loc(dctx.enclosingMethodLoc.file(), body->loc);
+    if (!loc.file().data(dctx.ctx).isRBI()) {
```
Comment (jez) [inline ast/desugar/Desugar.cc:170]:
```suggestion
    if (!dctx.enclosingMethodLoc.file().data(dctx.ctx).isRBI()) {
```

File: core/Loc.h (hunk 1 above)
Comment (jez) [reply core/Loc.h:33]: No it's fine to do it later or not do it.

**Pull request merged**